Call graph enhancements

smowton opened the pull request:
Two enhancements to the call graph, again both used in the security analyser but probably useful to others:

1. Optionally record callsites while building the graph, such that on seeing an A -> B edge in the callgraph it's easy to find out the calling instruction.
2. Support export to a `grapht`-derived data structure, such that standard graph algorithms can be used on the call graph.

smowton committed:
Call graph: optionally record callsites

This makes it easier for call-graph users to discover the particular instruction responsible
for a call-graph edge.

Ideally these would be recorded in-line with the map (i.e. change the graph type to
std::multimap<irep_idt, std::pair<locationt, irep_idt>>), but here we introduce an optional
auxiliary map to avoid breaking existing users.

smowton committed:
Call graph -> grapht transformation

Add export from call_grapht to grapht, such that generic graph algorithms
can be applied to the call graph.

polgreen commented:
Hi Chris, 

I've just had open source approval on some code I've written that rewrites the call_grapht class to inherit from grapht, because I wanted to use some of the base functions of grapht on the call graph. 
It also:
- computes the shortest path on the call graph and return the list of functions
- computes the functions reachable within N function calls from a given function.
 
I'll create the pull request for it later today, and it might make sense to try to combine the two together 

I think my pull request, would probably cover the need to export grapht from call_grapht, and using the grapht class makes it easier to get a callee from an edge (I think, without knowing how you are using that). The implementation of the invert function is, however, not as nice when using grapht.

smowton commented:
The callsites would be an edge attribute, i.e. don't use empty_edget. It should probably be optional if there's a risk of bloating the structure massively for programs with lots of parallel call-graph edges.

smowton commented:
(Also out of curiosity what is the open source approval process?)

polgreen commented:
Ah, I've used the nodes to store the function name irep_idt, and used empty_edget. I guess that storing the function name in the node and the fact that the nodes in a grapht have a vector of in edges, might eliminate the need for storing the caller name in the edge? Or storing the caller name in the edge could be used to eliminate the need to store the caller name in the node, but having both is probably not necessary.

Pull request here: 
https://github.com/diffblue/cbmc/pull/1509

It may need breaking down into several smaller pull requests.

smowton commented:
ACK, need to do more unification of @polgreen's changes and mine. Perhaps a Christmas holiday task for me :)

smowton committed:
Expose limited call graph via goto-instrument

smowton committed:
Slice global inits: use improved call graph

No behavioural change intended; should simply replace a graph walk
with one that the call graph code can now perform by itself.

smowton committed:
grapht: add get_reachable function

This simply performs a depth-first search of the graph starting from
a nominated node and returns a vector of reachable indices.

smowton committed:
Call graph: add constructors that only include reachable functions

These start from a nominated node and walk outwards, rather than build
the entire call-graph as usual.

smowton committed:
Add call graph helpers

These translate from function name to node ID and back, thus giving
more concise ways of running very simple graph algorithms over the
call graph.

smowton committed:
Add documentation to call graph

No functional changes

smowton committed:
Improve call graph unit tests

smowton commented:
@tautschnig @polgreen this is now ready for re-review. All features from @polgreen's branch have been integrated AFAIK. It now consists of quite a few pieces so I've divided it up into individual commits, which are best reviewed individually.

tautschnig commented:
@smowton I'm afraid doxygen isn't happy at the moment.

tautschnig commented on src/analyses/call_graph.cpp:
```diff
@@ -42,7 +45,12 @@ void call_grapht::add(
+        const irep_idt &callee=to_symbol_expr(function_expr).get_identifier();
+        add(function, callee);
+        if(collect_callsites)
+          callsites[{function, callee}].insert(i_it);
```
Why does this not use the newly introduced `add(caller, callee, callsite)` function?

smowton replied:
It does -- fixed in a later commit

tautschnig commented on src/analyses/call_graph.h:
```diff
@@ -20,23 +20,33 @@ Author: Daniel Kroening, kroening@kroening.com
-  explicit call_grapht(const goto_modelt &);
-  explicit call_grapht(const goto_functionst &);
+  explicit call_grapht(bool collect_callsites=false);
+  explicit call_grapht(const goto_modelt &, bool collect_callsites=false);
```
Safe to drop `explicit`

smowton replied:
Not so -- with a default argument there are one-arg versions of both constructors.

tautschnig commented on src/analyses/call_graph.cpp:
```diff
@@ -87,9 +87,67 @@ call_grapht call_grapht::get_inverted() const
 std::string call_grapht::format_callsites(const edget &edge) const
 {
-  PRECONDITION(collect_callsites, "collect_callsites should be enabled");
+  PRECONDITION(collect_callsites);
```
Nit pick: this change is not in the right commit.

tautschnig commented on src/util/graph.h:
```diff
@@ -469,6 +471,37 @@ void grapht<N>::visit_reachable(node_indext src)
 
+template<class N>
+std::vector<typename N::node_indext>
+grapht<N>::get_reachable(node_indext src, bool forwards) const
```
I'm not sure I like the fact that this basically does the same as `visit_reachable`, but just returns  the information in a different format -- yet neither is implemented in terms of the other.

tautschnig commented on src/analyses/call_graph_helpers.h:
```diff
@@ -0,0 +1,52 @@
+/// \param graph: call graph
+/// \param function: function to query
+/// \return vector of called functions
+std::vector<irep_idt> get_successors(
```
I'd suggest to use `std::set` or `std::unordered_set` -- otherwise it's not clear whether these will be unique or not. (They are not, but the ordering is completely arbitrary, which makes the result rather questionable.)

smowton commented:
@tautschnig all other changes applied. I implemented the non-const tagging visitor in terms of the const get-reachable, at the cost of some efficiency.

thk123 commented on src/analyses/call_graph.cpp:
```diff
@@ -84,11 +123,15 @@ void call_grapht::output(std::ostream &out) const
 
 void call_grapht::output_xml(std::ostream &out) const
 {
+  // Note I don't implement callsite output here; I'll leave that
```
Perhaps rather than a comment a `warning()` `if(collect_callsites)`

thk123 commented on unit/analyses/call_graph.cpp:
```diff
@@ -54,6 +54,7 @@ SCENARIO("call_graph",
     // {
     //    A();
     //    B();
+    //    B();
```
It seems dangerous to modify the unit test in the same commit since the behaviour of the unit test shouldn't change (since we are using the default value of not collecting callsites). (That is I think adding extra callsites is not a good idea as part of this PR. Extending the unit test to verify the callsites are indeed empty is very diligent :+1:

smowton replied:
The intent here is to highlight the difference between the multimap representation, which includes duplicate key-value pairs when  there are parallel edges, and the grapht representation which does not.

thk123 commented on src/analyses/call_graph.cpp:
```diff
@@ -35,9 +35,9 @@ call_grapht::call_grapht(
-  const goto_programt &body)
+static void forall_callsites(
+  const goto_programt &body,
+  std::function<void(goto_programt::const_targett, const irep_idt &)> callback)
```
Could you name `callback` something more descriptive e.g. `callsite_task`? And ideally add doxygen comments.

thk123 commented on src/analyses/call_graph.h:
```diff
@@ -25,6 +25,17 @@ class call_grapht
 
+  // These two constructors build a call graph restricted to functions
+  // reachable from the given root.
+  call_grapht(
```
We now have a lot of constructors - it might be worth considering a collection of named static functions instead? `create_callgraph_from_root_node`? Maybe even a builder pattern so the common parts can be setup first (e.g. `collect_callsites`).

thk123 commented on src/analyses/call_graph_helpers.cpp:
```diff
@@ -0,0 +1,62 @@
+
+#include "call_graph_helpers.h"
+
+static std::set<irep_idt> get_neighbours(
```
Nit: document (along with the rest of the functions in here). 

~Also, controversial suggestion, but `neighbours` doesn't mean much in the context of a callgraph, presumably it is the functions called by the specified function?~ Done using function successor and predecessor

thk123 replied:
I suppose actually it is either the called functions, or the callsites, depending on the value of `forwards` - this should definitely be documented.

thk123 replied:
Ah I see docs are in the header file. Personally prefer this, though I think we've been putting them in the source file - don't know if a decision has been reached @peterschrammel ?

thk123 commented on src/analyses/call_graph_helpers.cpp:
```diff
@@ -0,0 +1,62 @@
+  return result;
+}
+
+std::set<irep_idt> get_successors(
```
Consider calling this get_called_functions since we are specifically dealing with a call graph?

chrisr-diffblue reviewed (approved):
Apart from comments that have already been made by others, I'm happy to say this LGTM

smowton committed:
Rename call graph constructors
